client/servers/manager.go

```diff
@@ -0,0 +1,298 @@
+// Package servers provides an interface for choosing Servers to communicate
+// with from a Nomad Client perspective.  The package does not provide any API
+// guarantees and should be called only by `hashicorp/nomad`.
+package servers
+
+import (
+	"log"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+)
+
+const (
+	// clientRPCMinReuseDuration controls the minimum amount of time RPC
+	// queries are sent over an established connection to a single server
+	clientRPCMinReuseDuration = 5 * time.Minute
+
+	// Limit the number of new connections a server receives per second
+	// for connection rebalancing.  This limit caps the load caused by
+	// continual rebalancing efforts when a cluster is in equilibrium.  A
+	// lower value comes at the cost of increased recovery time after a
+	// partition.  This parameter begins to take effect when there are
+	// more than ~48K clients querying 5x servers or at lower server
+	// values when there is a partition.
+	//
+	// For example, in a 100K Nomad cluster with 5x servers, it will
+	// take ~5min for all servers to rebalance their connections.  If
+	// 99,995 agents are in the minority talking to only one server, it
+	// will take ~26min for all servers to rebalance.  A 10K cluster in
+	// the same scenario will take ~2.6min to rebalance.
+	newRebalanceConnsPerSecPerServer = 64
+)
+
+// Pinger is an interface for pinging a server to see if it is healthy.
+type Pinger interface {
+	Ping(addr net.Addr) error
+}
+
+// Server contains the address of a server and metadata that can be used for
+// choosing a server to contact.
+type Server struct {
+	// Addr is the resolved address of the server
+	Addr net.Addr
+	addr string
+	sync.Mutex
```

Comment:
What if we just used a NewServer func and always initialized the `addr`? I'm worried about what the Server copy in `Manager.GetServers` does to the Mutex. Alternatively GetServers could just manually copy Addr & DC into a new struct instead of copying the whole struct.

Reply:
@schmichael The two mutexes were put in place originally because Serf was handling the add/remove notifications and populating/replacing the server list independently of the server lists use via the RPC pool.  Basically: producer == serf/consul discovery, consumer == rpc pool.  This may be out of date, but that was the history.